sympy/solvers/tests/test_solveset.py

```diff
@@ -1886,6 +1886,56 @@ def test_raises_substitution():
     raises(TypeError, lambda: substitution([x**2 -1], x))
     raises(TypeError, lambda: substitution([x**2 -1], 1))
 
+
+def test_issue_21022():
+    # the answer is correct but the sequence is changing everytime
+    from sympy.core.sympify import sympify
+
+    eqs = [
+    'k-16',
+    'p-8',
+    'y*y+z*z-x*x',
+    'd - x + p',
+    'd*d+k*k-y*y',
+    'z*z-p*p-k*k',
+    'abc-efg',
+    ]
+    efg = Symbol('efg')
+    eqs = [sympify(x) for x in eqs]
+
+    syb = list(set.union(*[x.free_symbols for x in eqs]))
+    res = nonlinsolve(eqs, syb)
```

Comment:
What order are `syb` in when this passes for you? Locally this test fails:
```python
Traceback (most recent call last):
  File "main.py", line 5136, in <module>
    res = nonlinsolve(eqs, syb)
  File "main.py", line 5118, in nonlinsolve
    result = substitution(
  File "main.py", line 4774, in substitution
    new_result_real, solve_call1, cnd_call1 = _solve_using_known_values(
  File "main.py", line 4758, in _solve_using_known_values
    newresult, delete_res = _append_new_soln(
  File "main.py", line 4606, in _append_new_soln
    rnew[sym] = soln_imageset[sol]
KeyError: sqrt((z**2 - 16*z + 320)*(z**2 + 16*z + 320))/16
>>> syb
[x, k, p, d, y, abc, efg, z]
```

Reply:
Ahh...you say the sequence is changing every time. That is because you are passing `syb` in different order. There is no sorting requirement for sets. To make this canonical use `list(ordered(set.union(*[i.free_symbols for i in eqs])))`